redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/schemaJson.kt

```diff
@@ -0,0 +1,272 @@
+        tag = layoutModifier.tag,
+        type = layoutModifier.type.toJsonType(),
+        scopes = layoutModifier.scopes.map { it.toJsonType() },
+        properties = layoutModifier.properties.map {
+          JsonLayoutModifierProperty(
+            name = it.name,
+            type = it.type.toJsonType(),
+            defaultExpression = it.defaultExpression,
+          )
+        }
+      )
+    },
+    dependencies = dependencies.map {
+      JsonType(listOf(it.`package`, it.name))
+    },
+  )
+  val json = json.encodeToString(JsonSchemaV1.serializer(), schemaJson)
+
+  return EmbeddedSchema(
+    path = "${`package`.replace('.', '/')}/$name.json",
+    json = json,
+  )
+}
+
+private val json = Json {
+  prettyPrint = true
+  prettyPrintIndent = "\t"
+  classDiscriminator = "kind"
+}
+
+@Serializable
+private data class JsonType(
+  val names: List<String>,
+  val variance: Variance = Invariant,
+  val parameterTypes: List<JsonType> = emptyList(),
+  val nullable: Boolean = false,
+) {
+  enum class Variance {
+    @SerialName("in")
+    In,
+    Invariant,
+    @SerialName("out")
+    Out,
+  }
+
+  override fun toString(): String = buildString {
+    when (variance) {
+      In -> append("in ")
+      Invariant -> Unit // Nothing to do.
+      Out -> append("out ")
+    }
+    if (names[0] != "") {
+      append(names[0])
+      append('.')
+    }
+    names[1]
+    for (i in 2 until names.size) {
+      append('.')
+      append(names[i])
+    }
+    if (parameterTypes.isNotEmpty()) {
+      append('<')
+      parameterTypes.joinTo(this, separator = ", ")
+      append('>')
+    }
+    if (nullable) {
+      append('?')
+    }
+  }
+}
+
+private fun KClass<*>.toJsonType(): JsonType {
+  var qualifiedName = qualifiedName ?: throw AssertionError(this)
+  val simpleNames = ArrayDeque<String>()
+
+  var target: Class<*>? = java
+  while (target != null) {
+    target = target.enclosingClass
+
+    val dot = qualifiedName.lastIndexOf('.')
+    if (dot == -1) {
+      if (target != null) throw AssertionError(this) // More enclosing classes than dots.
+      simpleNames.addFirst(qualifiedName)
+      qualifiedName = ""
+    } else {
+      simpleNames.addFirst(qualifiedName.substring(dot + 1))
+      qualifiedName = qualifiedName.substring(0, dot)
+    }
+  }
+
+  simpleNames.addFirst(qualifiedName)
+
+  return JsonType(simpleNames)
+}
+
+private fun KType.toJsonType(): JsonType {
+  val kClass = classifier as? KClass<*>
+    ?: throw IllegalArgumentException("$this classifier must be a class")
+
+  return kClass.toJsonType().copy(
+    parameterTypes = arguments.map { it.toJsonType() },
+    nullable = isMarkedNullable,
+  )
+}
+
+private fun KTypeProjection.toJsonType(): JsonType {
+  val json = type?.toJsonType() ?: JsonType(names = listOf("*"))
+  return when (variance) {
+    null, INVARIANT -> json
+    IN -> json.copy(variance = In)
+    OUT -> json.copy(variance = Out)
+  }
+}
+
+@Serializable
+private data class JsonSchemaV1(
```

Comment:
I _think_ eventually the main `Schema` type will just be serializable and 90% of this copying will go away, but it's hard to see that far ahead.